### ISSUE TITLE
Formatting breaks psql variables

### ISSUE THREAD

**Issue by md384** [sqlfluff/sqlfluff#6686]
### Search before asking

- [x] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Postgres supports variables that can be substituted into SQL statements - see https://www.postgresql.org/docs/16/app-psql.html#APP-PSQL-INTERPOLATION.

When running sqlfluff with the postgres dialect and default settings these variables are reformatted into a broken state.

### Expected Behaviour

The following code should be unchanged

```sql
SELECT count(*)
FROM foo
WHERE status = :status;
```

### Observed Behaviour

A space is inserted between the `:` and variable name resulting in a syntax error.

```sql
SELECT count(*)
FROM foo
WHERE status =: status;
```

### How to reproduce

Format

```sql
SELECT count(*)
FROM foo
WHERE status = :status;
```

### Dialect

postgres

### Version

3.3.1

### Configuration

```
[sqlfluff]
dialect = postgres
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by md384** [comment]
Using the `colon_optional_quotes` for the `placeholder` templater works around this. However, there is an additional bug here when using a variable in an array slice, e.g. `some_arr[2 : :limit]` will get reformatted to `some_arr[2::limit]`.

```sql
\set position 2

select (ARRAY[1,2,3])[:position];
 array 
-------
     2
(1 row)

select (ARRAY[1,2,3])[1: :position];
 array 
-------
 {1,2}
(1 row)

select (ARRAY[1,2,3])[1::position];
ERROR:  syntax error at or near "position"
LINE 1: select (ARRAY[1,2,3])[1::position];
```

Perhaps there is a layout spacing rule that can workaround this but I haven't figured that out yet. This one feels like a bug though.

**Comment by madtibo** [comment]
Hi @md384 

I configured sqlfluff so it does not mess with the colons using the following configuration:

```
[sqlfluff:layout:type:colon]
spacing_before = any
spacing_after = any
```

But it would be nice to have a real fix  😍

**Comment by Rishi943** [comment]
Working on a fix for this - will open a PR shortly.
